File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left = 0
        right = 1
        max_profit = 0
        while right < len(prices):
            curr_profit = prices[right] - prices[left]
            if prices[left] < prices[right]:
                max_profit = max(curr_profit, max_profit)
            else:
                left = right
            right += 1
        return max_profit

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min_price_so_far = numeric_limits<int>::max(), max_profit = 0;
        for(const int price:prices){
            int max_profit_sell = price - min_price_so_far;
            max_profit = max(max_profit, max_profit_sell);
            min_price_so_far = min(min_price_so_far, price);
        }
        return max_profit;
    }
};